src/lib.rs

```diff
@@ -0,0 +1,631 @@
+            Err(_) => Err(RpcMapperError::UnexpectedError(String::from(
+                "Error while parsing Zenoh reply",
+            ))),
+        }
+    }
+}
+
+#[async_trait]
+impl RpcServer for ULinkZenoh {
+    async fn register_rpc_listener(
+        &self,
+        method: UUri,
+        listener: Box<dyn Fn(Result<UMessage, UStatus>) + Send + Sync + 'static>,
+    ) -> Result<String, UStatus> {
+        // Do the validation
+        UriValidator::validate(&method)
+            .map_err(|_| UStatus::fail_with_code(UCode::InvalidArgument, "Invalid topic"))?;
+
+        // Get Zenoh key
+        let zenoh_key = ULinkZenoh::to_zenoh_key_string(&method)?;
+        // Generate listener string for users to delete
+        let hashmap_key = format!(
+            "{}_{:X}",
+            zenoh_key,
+            self.callback_counter.fetch_add(1, Ordering::SeqCst)
+        );
+
+        let query_map = self.query_map.clone();
+        // Setup callback
+        let callback = move |query: Query| {
+            // Create UAttribute
+            let Some(attachment) = query.attachment() else {
+                listener(Err(UStatus::fail_with_code(
+                    UCode::Internal,
+                    "Unable to get attachment",
+                )));
+                return;
+            };
+            let Some(attribute) = attachment.get(&"uattributes".as_bytes()) else {
+                listener(Err(UStatus::fail_with_code(
+                    UCode::Internal,
+                    "Unable to get uattributes",
+                )));
+                return;
+            };
+            let u_attribute: UAttributes = if let Ok(attr) = Message::decode(&*attribute) {
+                attr
+            } else {
+                listener(Err(UStatus::fail_with_code(
+                    UCode::Internal,
+                    "Unable to decode attribute",
+                )));
+                return;
+            };
+            // Create UPayload
+            let u_payload = match query.value() {
+                Some(value) => {
+                    let Ok(encoding) = value.encoding.suffix().parse::<i32>() else {
+                        listener(Err(UStatus::fail_with_code(
+                            UCode::Internal,
+                            "Unable to get payload encoding",
+                        )));
+                        return;
+                    };
+                    UPayload {
+                        length: Some(0),
+                        format: encoding,
+                        data: Some(Data::Value(value.payload.contiguous().to_vec())),
+                    }
+                }
+                None => UPayload {
+                    length: Some(0),
+                    format: UPayloadFormat::UpayloadFormatUnspecified as i32,
+                    data: None,
+                },
+            };
+            // Create UMessage
+            let msg = UMessage {
+                source: Some(method.clone()),
```

Comment:
Following back up on this, @stevenhartley showed me [how this is accomplished in up-client-android-java](https://github.com/eclipse-uprotocol/up-client-android-java/pull/3/files#diff-fddb352c27884216476f4ee1bc524c0a823c41de369dad3f1cb038e4c4429643R214) last week, so I'll link it here in case you want to reference it.